### PR TITLE
Redesign the Go SDK resource/input/output system.

### PR DESCRIPTION
**Pull request by pgavlin**
The redesign is focused around providing better static typings and
improved ease-of-use for the Go SDK. Most of the redesign revolves
around three pivots:
- Strongly-typed inputs, especially for nested types
- Struct-based resource and invoke APIs
- Ease-of-use of Apply

### Strongly-typed inputs

Input is the type of a generic input value for a Pulumi resource.
This type is used in conjunction with Output to provide polymorphism
over strongly-typed input values.

The intended pattern for nested Pulumi value types is to define an
input interface and a plain, input, and output variant of the value
type that implement the input interface.

For example, given a nested Pulumi value type with the following shape:

```
type Nested struct {
    Foo int
    Bar string
}
```

We would define the following:

```
var nestedType = reflect.TypeOf((*Nested)(nil)).Elem()

type NestedInput interface {
    pulumi.Input

    ToNestedOutput() NestedOutput
    ToNestedOutputWithContext(context.Context) NestedOutput
}

type Nested struct {
    Foo int `pulumi:"foo"`
    Bar string `pulumi:"bar"`
}

type NestedInputValue struct {
    Foo pulumi.IntInput `pulumi:"foo"`
    Bar pulumi.StringInput `pulumi:"bar"`
}

func (NestedInputValue) ElementType() reflect.Type {
    return nestedType
}

func (v NestedInputValue) ToNestedOutput() NestedOutput {
    return pulumi.ToOutput(v).(NestedOutput)
}

func (v NestedInputValue) ToNestedOutputWithContext(ctx context.Context) NestedOutput {
    return pulumi.ToOutputWithContext(ctx, v).(NestedOutput)
}

type NestedOutput struct { *pulumi.OutputState }

func (NestedOutput) ElementType() reflect.Type {
    return nestedType
}

func (o NestedOutput) ToNestedOutput() NestedOutput {
    return o
}

func (o NestedOutput) ToNestedOutputWithContext(ctx context.Context) NestedOutput {
    return o
}

func (o NestedOutput) Foo() pulumi.IntOutput {
    return o.Apply(func (v Nested) int {
        return v.Foo
    }).(pulumi.IntOutput)
}

func (o NestedOutput) Bar() pulumi.StringOutput {
    return o.Apply(func (v Nested) string {
        return v.Bar
    }).(pulumi.StringOutput)
}
```

The SDK provides input and output types for primitives, arrays, and
maps.

### Struct-based APIs

Instead of providing expected output properties in the input map passed
to {Read,Register}Resource and returning the outputs as a map, the user
now passes a pointer to a struct that implements one of the Resource
interfaces and has appropriately typed and tagged fields that represent
its output properties.

For example, given a custom resource with an int-typed output "foo" and
a string-typed output "bar", we would define the following
CustomResource type:

```
type MyResource struct {
    pulumi.CustomResourceState

    Foo pulumi.IntOutput    `pulumi:"foo"`
    Bar pulumi.StringOutput `pulumi:"bar"`
}
```

And invoke RegisterResource like so:

```
var resource MyResource
err := ctx.RegisterResource(tok, name, props, &resource, opts...)
```

Invoke arguments and results are also provided via structs, but use
plain-old Go types for their fields:

```
type MyInvokeArgs struct {
    Foo int `pulumi:"foo"`
}

type MyInvokeResult struct {
    Bar string `pulumi:"bar"`
}

var result MyInvokeResult
err := ctx.Invoke(tok, MyInvokeArgs{Foo: 42}, &result, opts...)
```

### Ease-of-use of Apply

All `Apply` methods now accept an interface{} as the callback type.
The provided callback value must have one of the following signatures:

	func (v T) U
	func (v T) (U, error)
	func (ctx context.Context, v T) U
	func (ctx context.Context, v T) (U, error)

T must be assignable from the ElementType of the Output. If U is a type
that has a registered Output type, the result of the Apply will be the
corresponding Output type. Otherwise, the result of the Apply will be
AnyOutput.

Fixes https://github.com/pulumi/pulumi/issues/2149.
Fixes https://github.com/pulumi/pulumi/issues/3488.
Fixes https://github.com/pulumi/pulumi/issues/3487.
Fixes https://github.com/pulumi/pulumi-aws/issues/248.
Fixes https://github.com/pulumi/pulumi/issues/3492.
Fixes https://github.com/pulumi/pulumi/issues/3491.
Fixes https://github.com/pulumi/pulumi/issues/3562.

### CODE REVIEW

**Comment by pgavlin** [comment]
For a look at how this affects real-world resources, here's the diff in `pulumi-random`: https://github.com/pulumi/pulumi-random/compare/pgavlin/go

**Comment by pgavlin** [comment]
These changes still need some comments around the new input types, but they're otherwise mostly ready.

Note that the `ElementType` method on `Input` isn't yet being taken advantage of. We could use this for dynamic type-checking during serialization, and we will use this for the eventual implementation of `pulumi.MakeOutput` (or whatever we name it).

**Comment by pgavlin** [comment]
Oh, and here is an example of a program that uses a RandomId resource:

```
package main

import (
	"github.com/pulumi/pulumi-random/sdk/go/random"
	"github.com/pulumi/pulumi/sdk/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		randomId, err := random.NewRandomId(ctx, "random", &random.RandomIdArgs{
			ByteLength: pulumi.Int(16),
			Prefix:     pulumi.String("pulumi"),
		})
		if err != nil {
			return err
		}

		ctx.Export("randomId", randomId.ID)
		return nil
	})
}
```

`pulumi up`: https://asciinema.org/a/0Kxx0YceuTFwuQqiNpSQxc7IC

**Comment by lukehoban** [comment]
> Oh, and here is an example of a program that uses a RandomId resource:

Thanks - this is helpful.

Do you have an example of what you need to do to pass an `Output` as an `Input`?  

As a slightly more generalized thing - I was starting to try to port https://github.com/pulumi/examples/blob/7c28f4cf519c8e7910547ff6ef2b02fada432a88/aws-go-s3-folder/main.go to this new model to see what it feels like.  The few questions I had:
1. How do you pass `ID` from one resource as input to another?
2. How do `Asset`s work here - are they also Inputs?
3. Can `Map`s have strongly typed values, or will htey still need to be `map[string]interface{}`?

**Comment by pgavlin** [comment]
> As a slightly more generalized thing - I was starting to try to port https://github.com/pulumi/examples/blob/7c28f4cf519c8e7910547ff6ef2b02fada432a88/aws-go-s3-folder/main.go to this new model to see what it feels like. The few questions I had:

I figured it might be easier to show rather than tell, so here's the diff I've come up with for that example: https://github.com/pulumi/examples/compare/pgavlin/go

> How do you pass ID from one resource as input to another?

[See here](https://github.com/pulumi/examples/compare/pgavlin/go#diff-c1b46aa4de17652df6a40bfdfa63de06R39). Currently this requires a cast, but we could solve that by either generating union types or retyping the `ID` and `URN` outputs as `StringOutput`. There's really not much of a reason to strongly-type the former.

 > How do Assets work here - are they also Inputs?

They are, yes. For the example I gave, I cheated a little bit and made assets and archives assignable to `StringInput`. In the end, I expect we'll generate union input types for properties that accept assets/archives.

> Can Maps have strongly typed values, or will they still need to be map[string]interface{}?

The `Map` type in this PR is still `map[string]interface{}`, but we can generate more specific instantiations of this type as necessary. For example, I plan on adding a `StringMap` type that accepts `map[string]string`.

**Comment by joeduffy** [comment]
A couple minor questions just from looking at the S3 example (thank you for porting the examples! super helpful in understanding the overall impact):

* `Website: pulumi.Any(map[string]interface{}{`: I would have expected a `BucketWebsite` type, instead of a map. Is this just an alternative way of writing this in a less typed way?

* `Bucket:      pulumi.StringOutput(siteBucket.ID),`: It struck me as a little strange when reading that we're passing a `StringOutput` to an input ... should this be `pulumi.StringInput(siteBucket.ID)`? I feel like I must be just missing something fundamental here.

**Comment by pgavlin** [comment]
I've rebased this PR against master and updated the description with the current state of the changes.

I've also updated a few downstream repos:
- examples: https://github.com/pulumi/examples/compare/pgavlin/go
- pulumi-terraform-bridge: https://github.com/pulumi/pulumi-terraform-bridge/compare/pgavlin/goGen
- pulumi-random: https://github.com/pulumi/pulumi-random/compare/pgavlin/go
- pulumi-aws: https://github.com/pulumi/pulumi-aws/compare/pgavlin/go

PTAL.

File: sdk/go/pulumi/printf.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package pulumi
+
+import (
+	"fmt"
+	"io"
+)
+
+func allArgs(args []interface{}) ArrayOutput {
```
Comment (pgavlin) [inline sdk/go/pulumi/printf.go]: This is a bit of a wart that I'd like to fix up, but am happy to do so after the fact.

You would hope that `All(Any(args))` or `ToOutput(Any(args))` would work, but neither does. The implementation of these functions stops recursing and awaiting outputs when it hits a value that does not implement `Input`. In this case, that means we don't look through the `[]interface{}` to find outputs. Normally we stop at non-Input values because there isn't much we can do if we encounter an output--awaiting it likely won't produce a value we can assign to the location due to type differences (the location will be typed as an Output, and the awaited value will not)--but we should recurse through compound values looking for `interface{}` fields.

**Comment by pgavlin** [comment]
Here's what k8s looks like: https://github.com/pulumi/pulumi-kubernetes/compare/pgavlin/go

**Comment by pgavlin** [comment]
@lukehoban @EvanBoyle this is ready for review; PTAL. I'm working on adding a bit more testing for negative cases and the generated Apply/ToOutput/Index methods, but the core is there.

File: sdk/go/pulumi/context.go [hunk 2]
```diff
@@ -313,77 +435,45 @@ func (ctx *Context) RegisterResource(
...
-	providers map[string]ProviderResource
-}
+func (ctx *Context) RegisterComponentResource(
+	t, name string, resource ComponentResource, opts ...ResourceOption) error {
 
-// URN will resolve to the resource's URN after registration has completed.
-func (state *ResourceState) URN() URNOutput {
-	return state.urn
+	return ctx.RegisterResource(t, name, nil, resource, opts...)
 }
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply sdk/go/pulumi/context.go]: It's used to clarify the contract for registering component resources (i.e. no inputs, type must be a `ComponentResource`). Happy to remove this if it seems better.

**Comment by pgavlin** [comment]
> Could we add all the issues this resolves to the description?

Will do.

> What is the plan for pushing out the impact of this change to all other repos? That is - what is on the critical path to everything at head working cleanly again for Go? Can we make sure we have a P1 issue tracking that critical path work so we can drive it to completion quickly? (while also tracking all the other work we want to do to continue to improve Go support).

I'll break this down per-repo-class.

#### TF bridge & providers
There are two routes we can take to get the codegen right:
1. Merge https://github.com/pulumi/pulumi/pull/3749 and the Go portion of the changes from https://github.com/pulumi/pulumi-terraform-bridge/compare/pgavlin/schema
2. I can backport the Go codegen changes from https://github.com/pulumi/pulumi/pull/3749 onto the TF bridge IR

Once the codegen is adjusted, the bridge just needs to be rolled out to the dependent repos.

#### Kubernetes provider

I have some changes in progress to bring up k8s codegen, but they still need a little bit of work before they're ready. If we merge https://github.com/pulumi/pulumi/pull/3749, we could use that package's code generator for the k8s Go SDK.

#### Examples

We do not have many Go examples, and I've ported them all in the past. Just need to get changes into a branch and get them merged.

**Comment by rogercoll** [comment]
Is this issue closed? I couldn't find any example to convert a `StringOutput` to `String`. 
I faced that problem while specifying the `KeyName` of an ec2 instance:

`cannot convert mykey.KeyName (type pulumi.StringOutput) to type pulumi.String`

In this case the user just specifies the name of the key to be created. I could get this name but I reckon I would be great to be able to convert it.

**Comment by pgavlin** [comment]
> I couldn't find any example to convert a StringOutput to String.
I faced that problem while specifying the KeyName of an ec2 instance:

It is not possible to convert a `StringOutput` to a `String`. A `StringOutput` carries additional information (e.g. the resources the value depends on) and may not be promptly available, so this conversion is not safe. If you need to transform the value of a `StringOutput`, you can use `ApplyT` to do so: https://pkg.go.dev/github.com/pulumi/pulumi/sdk@v1.14.0/go/pulumi?tab=doc#OutputState.ApplyT
